Library/Homebrew/unpack_strategy.rb

```diff
@@ -0,0 +1,241 @@
+class UnpackStrategy
+  def self.from(download:); end
+
+  def self.magic_number(path)
+    File.binread(path, 252)
+  end
+
+  def self.detect(path)
+    magic_number = magic_number(path)
+
+    strategies = [
+      JarUnpackStrategy,
+      ZipUnpackStrategy,
+      XarUnpackStrategy,
+      XzUnpackStrategy,
+      TarUnpackStrategy,
+      Bzip2UnpackStrategy,
+      GzipUnpackStrategy,
+      LzipUnpackStrategy,
+      P7ZipUnpackStrategy,
+      RarUnpackStrategy,
+      LhaUnpackStrategy,
+    ]
+
+    strategy = strategies.detect do |s|
+      s.can_extract?(path: path, magic_number: magic_number)
+    end
+
+    # This is so that bad tarballs and zips produce good error messages.
+    strategy ||= case path.extname
+    when ".tar.gz", ".tgz", ".tar.bz2", ".tbz"
+      TarUnpackStrategy
+    when ".zip"
+      ZipUnpackStrategy
+    else
+      UncompressedUnpackStrategy
+    end
+
+    strategy.new(path, magic_number: magic_number)
+  end
+
+  attr_reader :path, :magic_number
+  private :magic_number
+
+  def initialize(path, magic_number:)
+    @path = Pathname(path).expand_path
+    @magic_number = magic_number || self.class.magic_number(path)
+  end
+
+  def extract(to: nil)
+    unpack_dir = Pathname(to || Dir.pwd).expand_path
+    unpack_dir.mkpath
+    extract_to_dir(unpack_dir)
+  end
+
+  def pipe_to_tar(tool, unpack_dir)
+    Utils.popen_read(tool, "-dc", path) do |rd|
+      Utils.popen_write("tar", "-x", "-f", "-", "-C", unpack_dir) do |wr|
+        buf = ""
+        wr.write(buf) while rd.read(16384, buf)
+      end
+    end
+  end
+  private :pipe_to_tar
```

Comment:
Can you order these so you can have a single `private`?

Reply:
I'm trying to follow the new RuboCop rule (which we disabled for now because of false positives). Basically, the `private` should be next to the method, so it is much easier to tell if a method is private or not.

Reply:
Does that rule have an autocorrect? If not I think it's going to prove too painful to implement this consistently across the codebase (and I think remaining consistent for now would be better).